Fix freezes due to large circular arc radius

Naxesss opened the pull request:
(This is my first PR here so do let me know if I'm doing anything incorrectly.)

This resolves https://github.com/ppy/osu/issues/5538 and the freezing/crashing portion of https://github.com/ppy/osu/issues/12047. @bdach outlines this issue pretty well in both of those.

We could potentially fix this in [`PathApproximator`'s `ApproximateCircularArc` in the osu!framework](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Utils/PathApproximator.cs#L165) instead, but this poses the risk of arbitrarily limiting the radius of circular arcs based on osu's playfield dimensions, so I think we should solve this in the client.

It is also in the client where the performance bottleneck seems to occur; placing a breakpoint after the `PathApproximator.ApproximateCircularArc` line revealed that it's hit almost instantly (compared to how long the freezes last) even for cases with 4k+ subpath points. I noticed split-second freezing after ~500 points in-game, so 400 should give a decent margin while still allowing large circles to be made.

![](https://i.imgur.com/G2DuwXl.gif)

I am a bit unsure of where to put the test. I was considering splitting it off into a `TestSceneSliderPlacementEdgeCaseBlueprint` class, as the tests within `TestSceneSliderPlacementBlueprint` are mostly regular use cases. However, if I do that then the `assertLength` etc functions would need to be duplicated unless moved to another class and inherited by both. I figured it'd be worth asking about that before possibly wasting effort if the current is ok.

- [x] depends on https://github.com/ppy/osu-framework/pull/4313

Naxesss committed:
Add linear circular arc test

Naxesss committed:
Fix freeze due to large circular arc radius

Seems to stem from the osu!framework's PathApproximator not catching a few edge cases wherein the radius approaches infinity.

bdach commented:
A few things:

> We could potentially fix this in `PathApproximator`'s `ApproximateCircularArc` in the osu!framework instead

I agree with not doing this, and it was also the consensus from previous discussions on how to approach this that this should be a game-local thing.

> It is also in the client where the performance bottleneck seems to occur

To clarify this, the bottleneck is not computing the slider path, it's drawing it. For the slider to be drawn, FBOs are allocated over at the framework side; in the regressive cases the radius of the slider increases by a lot, causing frame drops because of the sheer amount of time required to render it.

---

While predicating the cutoff on the number of points in the approximation is certainly a way to check whether the slider is too large, it's not exactly what stable does. For reference, stable does this:

```csharp
                    OsuMathHelper.CircleThroughPoints(a, b, c, out centre, out radius, out t_initial, out t_final);

                    if (GameBase.Mode == OsuModes.Edit)
                    {
                        // Find the circle's bounding box
                        // Don't fall back to bezier if the slider is already placed and
                        // eg. the user rotates it.
                        // https://osu.ppy.sh/forum/t/116612
                        osu.Helpers.BoundingBox box = new Helpers.BoundingBox(OsuMathHelper.CirclePoint(centre, radius, t_initial), OsuMathHelper.CirclePoint(centre, radius, t_final));
                        double t_low = Math.Min(t_initial, t_final);
                        double t_high = Math.Max(t_initial, t_final);
                        int q = (int)(t_low * 2.0d / Math.PI + 1);
                        while (q * 0.5f * Math.PI < t_high)
                        {
                            box.Add(OsuMathHelper.CirclePoint(centre, radius, q * 0.5f * Math.PI));
                            q++;
                        }

                        if ((placingPending && ((float)box.Width > 512 || (float)box.Height > 384))
                            || ((float)box.Width > 640 || (float)box.Height > 640))
                            goto case CurveTypes.Bezier;
                    }
```

In other words, this Bezier conversion logic is only active in the editor. The change as done here would affect all sliders, including the ones in gameplay, therefore possibly breaking maps that bypassed that editor limitation (via manual editing or otherwise), so I'm not sure we should proceed with this fix.

I had looked at this recently, and attempted to make this logic local to the editor by placing that bounding box limiting logic in `PathControlPointPiece` if I recall correctly, but it was not working all that well due to reasons that I haven't managed to look into too much yet.

Naxesss commented:
> In other words, this Bezier conversion logic is only active in the editor, and only in specific cases. The change as done here would affect all sliders, including the ones in gameplay, therefore possibly breaking maps that bypassed that editor limitation (via manual editing or otherwise), so I'm not sure we should proceed with this fix.

That is a good point, I didn't realize it was limited to the editor. In that case this specific fix could possibly break aspire maps, so will close this.

peppy commented:
We might still be able to use an approach like this but apply it closer to the editor's code. I haven't looked into it in detail, but for instance it may be possible to replace the `SliderPath` with a custom `EditSliderPath`, or something similar via `DrawableOsuEditRuleset`.

Either way, I'd probably keep this PR open for the time being until we decide that it's 100% not a valid direction to follow.

bdach commented:
`SliderPath` is a hitobject-level entity, not a drawable entity. So it'd probably have to be done at a deeper level than at the drawable ruleset (at conversion time or something?)

peppy commented:
Even there, we so have control over this via `SliderPlacementBlueprint`, but that does seem like it'd bee too deep.

Another option may be running the calculation once locally in `PathControlPointVisualiser` before actually applying the changes to the `Slider` itself, to catch the excessive case.

frenzibyte commented:
Also, as an aside, do not open pull requests off of a `master` branch, give the [contributing guidelines](https://github.com/ppy/osu/blob/master/CONTRIBUTING.md#i-would-like-to-submit-a-pull-request) a read if you haven't already.

Naxesss commented:
Alright reopening.

@frenzibyte I did read them earlier but forgot about that part. Thanks though, will keep in mind for next time (assuming I can't change branches of this PR).

I've looked around a few of the other editor-related classes and as far as I've gathered, `PathControlPointVisualizer` doesn't currently seem to be receptive to moving points, only selecting, creating, and deleting. `PathControlPointPiece` has `OnDrag`, which catches moving nodes, but not the preview when initially placing the slider. So that's a problem.

Having played around with `OnDrag` a bit, however, it seems possible to isolate the control points responsible for creating a circular arc, and then running their positions through the bezier approximation to then decide whether to temporarily change the PathType, like this (note the control point changing colour):

![](https://i.imgur.com/0pBO5Tz.gif)

It's not a very elegant solution, though, as this calculates the bezier approximation twice for every path (once to know which path type to use and once in `SliderPath`) - can probably reuse/cache that somehow. Also currently, if you let go of the slider when it is in bezier mode, it stays that way (the original `PathType` is stored only for each drag), unlike in stable.

bdach commented:
@Naxesss I think that gif looks like a good start, and I wouldn't really put much mind to the fact that releasing in an invalid position will convert the node to Bezier myself. Especially given it's an area in which stable would essentially silently lie to the user about what it's doing in the editor, and reproducing stable's behaviour would probably result in way too much complexity for what it's worth.

That said if we do go in this direction, it would be probably best to also disable the perfect curve type in the control point's context menu, which sounds like it will not be easy or pretty...

peppy commented:
Having the colour / type change is definitely what we want, rather than hiding it behind internal switches like stable does. It should probably only "commit" the curve type when letting go of the drag, so the user can return to a sane `perfect` curve if that is their intention.

smoogipoo commented:
This may be fine as a temporary solution, but I don't think it's viable in the long term. My primary concern is, with this being done at `SliderPath` level, breaking maps that take advantage of these mechanics (e.g. https://osu.ppy.sh/beatmapsets/756794#osu/1605148), so I'll at least have to run a diffcalc of these changes.

Furthermore, this is still going to wreck systems that have lower VRAM than yours:
![image](https://user-images.githubusercontent.com/1329837/111978093-e4f27a80-8b46-11eb-815d-e75237ba162c.png)
So the better solution is to limit FBO sizes, but since that hasn't been done yet this temporary change is fine. Just keep in mind that these sort of scenarios can happen in-game even with your heuristic - 200MB can be quite large on some (especially integrated/mobile) systems.

As an aside, all stable does is limit the bounding box to 640x480 before changing to bezier. If that can be done instead directly in the blueprints rather than `SliderPath` itself, I'd highly prefer it.

Naxesss committed:
Undo subpath limiting

Naxesss committed:
Add PointsInSegment method

Naxesss committed:
Add path type update logic

Only attempts to change points to bezier if points in the slider are modified.

Naxesss committed:
Maintain path type when dragging/placing

Naxesss committed:
Add path type test

Naxesss committed:
Add path type recovery test

Naxesss commented:
> Maybe it is worth it to come up with a slightly different test for this scenario (e.g. one that relies on the dot product/cosine of the two vectors spanning the arc) to avoid computing the path every time?

One alternative that works well is this:
```csharp
if (pathType == PathType.PerfectCurve)
{
    Vector2[] points = pointsInSegment.Select(point => point.Position.Value).ToArray();
    if (points.Length < 3)
        return;

    Vector2 a = points[0];
    Vector2 b = points[1];
    Vector2 c = points[2];

    float maxLength = points.Max(p => p.Length);

    Vector2 normA = new Vector2(a.X / maxLength, a.Y / maxLength);
    Vector2 normB = new Vector2(b.X / maxLength, b.Y / maxLength);
    Vector2 normC = new Vector2(c.X / maxLength, c.Y / maxLength);

    float det = (normA.X - normB.X) * (normB.Y - normC.Y) - (normB.X - normC.X) * (normA.Y - normB.Y);

    float acSq = (a - c).LengthSquared;
    float abSq = (a - b).LengthSquared;
    float bcSq = (b - c).LengthSquared;

    // Exterior = curve wraps around the long way between end-points
    // Exterior bottleneck is drawing-related, interior bottleneck is approximation-related,
    // where the latter is much faster, hence differing thresholds
    bool exterior = abSq > acSq || bcSq > acSq;
    float threshold = exterior ? 0.05f : 0.001f;

    if (Math.Abs(det) < threshold)
        pointsInSegment[0].Type.Value = PathType.Bezier;
}
```
Which catches both cases of freezing:
- Middle point being beside the end-points, where as @bdach explained earlier we quickly run out of memory due to FBO allocation.
- Middle point being in between the end-points, where not much is rendered so I'd assume must be approximation-related (possibly the list of subpaths becoming very large as the radius approaches infinity).

---

I've tried various ways of applying this, and the only method I've found that works for all slider path changes (both when initially placing and when moving already placed control points), is to subscribe to control point changes in `SliderPath`. This way only *editing* is affected, while *gameplay* is kept the same. So the map @smoogipoo linked still works like in stable, for example.

However by doing it here instead of in `PathControlPointPiece`, we lose the context of the drag, which means every time it changes the path to Bezier, it "commits" as @peppy mentioned shouldn't happen. A workaround for this is to constantly try to maintain the path type while dragging/placing (where `SliderPath` updates it last), giving us this behaviour as desired:

![](https://i.imgur.com/XoYpGsE.gif)

![](https://i.imgur.com/lxJ1SuJ.gif)

(I did also look at the blueprint classes, but they are about placing new objects, not modifying existing ones from what I can tell. So that would at best only catch the case in the latter gif, if I'm not mistaken.)

Will push the code so we can take a look at it. I added a `PointsInSegment` method in `SliderPath`, as I noticed a very similar implementation was done by @bdach [here on another issue](https://github.com/ppy/osu/issues/11885#issuecomment-803643564).

Naxesss committed:
Fix newlines/spaces

Naxesss committed:
Fix null checks

Naxesss committed:
Fix missing newline

Naxesss committed:
Invert if statement

Naxesss committed:
Improve fallback conditions

It's possible to create a `PerfectCurve` type path with more than 3 points currently, so this accounts for that.

Naxesss committed:
Separate condition logic from math logic

Naxesss committed:
Add method documentation

Naxesss committed:
Fix inverted return statement

Forgot to run tests, all passing now.

bdach commented:
The gifs look really good, but I'm not super sure still on placing that logic in `SliderPath`. Feels too close for comfort, if you will.

I'll see what everyone else thinks first, but a possible avenue that would make this nicer for me would be to have some sort of `SliderEditorManager` (bad name, just to illustrate purpose), shared by placement and selection blueprints, that would operate on a basis of something like:

```
backup slider state from before change

while change is not committed:
	check current path state
	if path invalid
		bring path to correct state (by rolling back, changing node types, etc.)

on commit, discard backed up state (because it was either re-applied, or is no longer relevant)
```

It would make for one central place for loosely enforcing the invariants that a slider path must adhere to, rather than spreading them all around the place or putting them somewhere where they may possibly impact gameplay if someone is not careful.

peppy commented:
The code is pretty sound, but can agree that we *probably* don't want it in `SliderPath`, although I would rather it be there than having this issue remain broken.

I think it may be feasible to just move the latest code you have quite verbatim into `PathControlPointVisualiser`, using the bindable flow there. Give it a try and let us know if you run into any difficulties.

Naxesss committed:
Update PointsInSegment when adding/removing points

There was a bug where if you created a slider, moved the last point, and then added a point such that it became a PerfectCurve, it would fail to recover after becoming a Bezier.

This fixes that.

Naxesss committed:
Simplify path type maintenance when dragging

Naxesss committed:
Move path type correction

This is better because `PathControlPointVisualizer` is local to the editor, meaning there is no chance that this could affect gameplay.

Naxesss committed:
Add control point dragging tests

Naxesss committed:
Remove unnecessary length asserts

We don't actually care about the length (as this isn't what we're testing), just the type of the slider.

Naxesss committed:
Fix coordinates

Naxesss committed:
Shorten recovery test name

Naxesss committed:
Add exterior arc test

Naxesss commented:
Alright moved to `PathControlPointVisualizer`, so should be entirely local to the editor now. Also added a few more tests (including ones for dragging control points) to ensure this sticks.

peppy commented:
The implementation is looking a lot better now, but the issue mentioned [in this comment](https://github.com/ppy/osu/pull/12122#issuecomment-803965298) still stands: the limiting needs to be more stringent.

![20210324 132752 (dotnet)](https://user-images.githubusercontent.com/191335/112255007-d326e980-8ca4-11eb-8972-83e1de639d1a.png)

As it stands, I can still make a slider which uses 200mb of RAM, and double that when selecting the blueprint itself. Rather than getting to a point where the slider is 8k pixels, the limiting should kick in closer to when the slider exceeds the playfield size.

Naxesss commented:
Just to be clear, would making the threshold more strict in `validCircularArc` solve this? For example upping it from 0.05 to 0.4:

![](https://i.imgur.com/jlmoJ30.gif)

peppy commented:
Yep, should be fine. It should be high enough that the resultant slider is larger than the visible screen area so it doesn't falsely trigger. What you have in the screen capture seems a lot more appropriate.

Naxesss committed:
Add circular arc size tests

Naxesss committed:
Increase exterior threshold

Naxesss committed:
Fix broken control point drag test

Broken for 2 reasons:
- Assert checks the wrong control point.
- The exterior arc is now too big.

This fixes both.

Naxesss commented:
Oh actually increasing that threshold might not be a good idea since it doesn't really care about size, so that would also stop smaller circles from being made... I'll try something else.

Naxesss committed:
Update large arc tests

Should now be more robust and readable.

Naxesss committed:
Add complete arc test

Ensures we can still make smaller circles properly.

Naxesss committed:
Add bounding box limit

Naxesss commented:
@peppy do I keep all the math logic inside `PathControlPointVisualizer`? Seems a bit strange to have so much specific math logic there, but not sure where else to put it. Figured perhaps the slider itself kept track of its bounding box but that seems to be the drawable, and unsure how/if I should access that.

bdach commented:
It does feel wrong to have essentially a copy of the path approximating logic from framework in here. I see two moderately easy ways out of this:

* either to move this bounding-box computing method to framework (reusing as much of the approximator code there as possible),
* or to create a copy of the existing path, apply the user changes, and then compute the entire path (see [this old discord message I dug up](https://discord.com/channels/188630481301012481/188630652340404224/775341077481783307) from when I originally tried this and gave up for reasons I don't entirely recall).

The latter is easier because it won't require any framework changes, but it will be a lot more wasteful wrt allocations.

Naxesss committed:
Remove determinant limit

This has since been added into the framework through https://github.com/ppy/osu-framework/pull/4302

Naxesss committed:
Move bounding box logic to framework

FamousPig commented:
I stumbled across an issue. The curve type isn't reverted as perhaps expected when scaling.

https://user-images.githubusercontent.com/22643729/112827253-5de55b00-908e-11eb-9ab6-9eb463bef275.mp4

Naxesss committed:
Add scaling path type recovery

bdach commented on osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs:
```diff
@@ -142,6 +148,23 @@ public void OnReleased(PlatformAction action)
+
+                ReadOnlySpan<Vector2> points = slider.Path.PointsInSegment(segmentStartPoint).Select(p => p.Position.Value).ToArray();
+                RectangleF boundingBox = PathApproximator.CircularArcBoundingBox(points);
+                if (points.Length == 3 && boundingBox.Area >= 640 * 480)
```
An area check seems a little wrong in this context. For instance, a 20x15359 slider would still theoretically pass it. It's probably harder to abuse this than that random example, given that the bounding box is based on an arc, but I'd say checking individual dimensions would be preferable.

bdach commented on osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs:
```diff
@@ -91,6 +93,8 @@ private void onControlPointsChanged(object sender, NotifyCollectionChangedEventA
 
                         Connections.Add(new PathControlPointConnectionPiece(slider, e.NewStartingIndex + i));
+
+                        point.Changed += updatePathTypes;
```
This subscription seems leaky (likely holds a reference to visualiser instances after they've been disposed). Probably need to double-check if this is safe and unsubscribe from this event for all points in `Dispose()` if not.

Naxesss replied:
Binding to `point.Position` breaks the (probably desired) behaviour seen in the bottom gif of [this post](https://github.com/ppy/osu/pull/12122#issuecomment-804190126).

Ended up creating a local bound copy of the `Path.Version` bindable inside `PathControlPointPiece`. This way I can also remove the loop, as all control points will now call this when the path updates.

Naxesss committed:
Update/remove determinant tests

We now only change the path type based on the bounding box.

If the control points are too linear, the framework now handles the fallback to Bezier.

Naxesss committed:
Move updatePathTypes to PathControlPointPiece

Here we produce a local bound copy of the path version, and bind it to update the path type.

This way, if the path version updates (i.e. any control point changes type or position), we check that all control points have a well-defined path.

Additionally, if the control point piece is disposed of, the GB should also swoop up the subscription because of the local bound copy.

Naxesss committed:
Check current point instead of start point

Since each control point will call this when the path updates, the previous would correct the start segment 3 times instead of just once.

This fixes that.

Naxesss committed:
Undo public -> internal for `PathControlPoint.Changed`

No longer used.

Naxesss committed:
Fix broken test case

Seems this technically works, but only because of the edge case of being entirely linear, which the framework catches.

This fixes that.

peppy commented:
This is going to need some minor conflict resolution. Let me know if you're unsure of the correct merge path.

Naxesss committed:
Merge branch 'master' into circular-arc-freeze

Naxesss committed:
Remove unused blueprint variable

Naxesss committed:
Add margin to large segment test

Test ran fine on my end, but apparently not on the CI. This should make results a bit more consistent, hopefully.

peppy commented:
Looks like this doesn't guard correctly for perfect curve sections in a multi-segmented slider:

https://user-images.githubusercontent.com/191335/113860819-70c00400-97e1-11eb-98de-048b9e714478.mp4

Naxesss committed:
Add regression test for type changes

Naxesss committed:
Track path type changes for `PointsInSegment`

Naxesss commented:
Cause seems to be that `PointsInSegment` doesn't update when changing control point types in the slider.
Solved that by tracking the path types, using `pathTypes` to hold the bound copy references.

Also added a regression test for this case.

Before:
![image](https://user-images.githubusercontent.com/30292137/113890996-f5e8f000-97c4-11eb-9a93-d9f696adf892.png)

After:
![image](https://user-images.githubusercontent.com/30292137/113891154-19ac3600-97c5-11eb-9576-befc3f871a69.png)

Naxesss committed:
Convert `pathTypes` to local variable

Not entirely sure what holds the reference to pathTypes now (the binding to`slider.Path.ControlPoints` maybe?), but this does seem to work still.

Naxesss committed:
Simplify slider path bindings

Adds a slight performance overhead, but solves the memory leak and makes the code much easier to follow.

peppy committed:
Merge branch 'master' into circular-arc-freeze

peppy reviewed (approved):
This is looking and feeling pretty correct now. Good stuff!
